### DISCUSSION TITLE
Bot Learning Match to Match

### DISCUSSION THREAD

**Discussion by braftery77** [SebLague/Chess-Challenge#217] [General]
So one of my bots I was thinking about making involves learning over time from match to match. However if the bot is instantiated for each match then this wouldn't really work. Since the idea of Think() is to have a board state and then calculate the next best move, I don't think that it would need to be wiped between matches and can be made to learn as it plays games. Not saying this strategy would be the best bot, but I think it would be interesting to try out.

**Comment by JoseBritto** [comment]
You can store data in static variables if you don't want the data to be reset every match. But imo you would have to store it to some file and before submitting store the learned data in a string or something similar as you can't use IO functions as per the rules.

**Reply by braftery77** [reply]
This is a good idea. Didn't think about the static variable sticking around.

**Reply by godammit** [reply]
hey im new to C# (i usually do python, php and sql) and i thought this project would be good to learn C#, since my plan is to create an AI that learns over time. i'd like to store its weights and bias' so if it discovers a better way of playing (by adjusting the neural network every time it plays, and if its worse it reverts back to the former network.) but i've never heard of "static" variables. is it just the "constant" variables?

**Comment by mcthouacbb** [comment]
Do you mean learning before the final tournament and then submitting that data with the MyBot.cs file, or learning from matches while running the tournament?

**Reply by braftery77** [reply]
I meant for learning while running the tournament

**Reply by JoseBritto** [reply]
> I meant for learning while running the tournament

Tbh, I don't think it will play out very well in your favour in that case. We don't know how Sebastian is gonna run the tournament, maybe he might spawn all submissions as objects and make them play against each other without the program ever quitting, in which case this might play in your favour (assuming your bot has a good enough strategy from the start to iterate on so that it won't get weeded out at the very beginning). But in the more likely case the bot is stopped at least once while running the tournament (maybe like after a round or something similar) or even worse it is restarted every match, the bot would have to start from scratch every single time.

But if you can achieve this without compromising much, it could give your bot an edge.

Similar thought that came to my mind, it would be super cool if a bot could find the weak spot in the strategy of the opponent before the current match ends and adapt accordingly, but I have no idea how to even begin to implement something like that let alone how to make it fit into this tiny program.
